Use Service ClusterIPs as MC Service's Endpoints

luolanzone opened the pull request:
This PR is on top of #3603, only top commit is for Endpoint change. 

Use Service ClusterIPs instead of Pod IPs as MC Service's Endpoints.
The ServiceExport controller will only watch ServiceExport and
Service events, and wrap Service's ClusterIPs into a new Endpoint kind of
ResourceExport.

Signed-off-by: Lan Luo <luola@vmware.com>

luolanzone commented:
/test-multicluster-dataplane-e2e

luolanzone commented:
/test-multicluster-dataplane-e2e

luolanzone commented:
/test-multicluster-dataplane-e2e

luolanzone commented:
/test-multicluster-dataplane-e2e

luolanzone commented:
/test-multicluster-dataplane-e2e

luolanzone commented:
/test-multicluster-dataplane-e2e

jianjuns commented on multicluster/controllers/multicluster/common/helper.go:
```diff
@@ -82,3 +82,29 @@ func FilterEndpointSubsets(subsets []corev1.EndpointSubset) []corev1.EndpointSub
 	return newSubsets
 }
+
+func GetSubsets(svc *corev1.Service) corev1.EndpointSubset {
```
Name it GetServiceEndpointSubnet or GetEndpointSubnet

jianjuns commented on multicluster/controllers/multicluster/common/helper.go:
```diff
@@ -82,3 +82,29 @@ func FilterEndpointSubsets(subsets []corev1.EndpointSubset) []corev1.EndpointSub
+}
+
+// PortsConverter will convert Service's port to EndpointPort
+func PortsConverter(ports []corev1.ServicePort) []corev1.EndpointPort {
```
ConvertServiceEndpointPorts or GetServiceEndpointPorts or GetEndpointPorts

luolanzone replied:
done

jianjuns commented on multicluster/controllers/multicluster/serviceexport_controller.go:
```diff
@@ -257,20 +235,7 @@ func (r *ServiceExportReconciler) Reconcile(ctx context.Context, req ctrl.Reques
+	ep.Subsets = []corev1.EndpointSubset{common.GetSubsets(svc)}
 	if epInstalled {
 		installedEp := epObj.(*epInfo)
 		if apiequality.Semantic.DeepEqual(getEndPointsPorts(ep), installedEp.ports) &&
```
Should we change how to detect no EP change? As we just need to check if Service ClusterIP/port changed, we no long need to check individual Endpoints?

luolanzone replied:
done, removed Endpoint comparison.

luolanzone commented:
/test-multicluster-dataplane-e2e

luolanzone commented:
/test-multicluster-dataplane-e2e

luolanzone commented:
/test-multicluster-dataplane-e2e

jianjuns reviewed (dismissed):
LGTM

luolanzone commented:
/test-multicluster-dataplane-e2e

luolanzone commented:
/test-multicluster-dataplane-e2e

luolanzone commented:
/test-multicluster-dataplane-e2e

luolanzone commented:
/test-multicluster-dataplane-e2e

luolanzone commented:
/test-multicluster-dataplane-e2e

luolanzone commented:
/test-multicluster-dataplane-e2e

luolanzone commented:
/test-multicluster-dataplane-e2e

luolanzone commented:
/test-multicluster-dataplane-e2e

luolanzone commented:
/test-multicluster-dataplane-e2e

luolanzone committed:
Add Multi-cluster feature in Agent

* Add a new feature gate `Multicluster` and configs in antrea-agent.conf, and a few extra items in antrea-agent
  cluster role including access to `Gateway` and `ClusterInfoImport`.
* Add a controller for Gateway Nodes to watch Gateway and ClusterInfoImport's
events. It will set up a few openflow rules to forward cross-cluster traffic to remote Gateway Nodes.
    * Add a classification rule for cross-cluster traffic with global multicluster virtual
    MAC `aa:bb:cc:dd:ee:f0`. A sample is like below:
    ```
     table=Classifier, priority=210,in_port="antrea-tun0",dl_dst=aa:bb:cc:dd:ee:f0 actions=load:0x1->NXM_NX_REG0[0..3],load:0x1->NXM_NX_REG0[9],resubmit(,SNATConntrackZone)
    ```
    * Add a de-SNAT rule for reply packets. A sample is like below (the destination IP is local Gateway IP):
    ```
    table=SNATConntrackZone, priority=210,ip,nw_dst=10.176.25.59 actions=ct(table=ConntrackZone,zone=65530,nat)
    ```
    * Add a rule in ConntrackZone for reply packets if it's traffic from regular Node through checking ct_mark.
    The request packets from regular Node is not committed in Gateway Node, so we need to skip ConntrackStateTable.
    ```
    table=ConntrackZone, priority=210,ct_mark=0x1/0xf,ip,nw_src=10.96.0.0/12 actions=resubmit(,PreRoutingClassifier)
    ```
    * Add a rule in L3Forwarding table for cross-cluster request packets that modifies
    the destination MAC to global multicluster virtual MAC. A sample is like below (the destination CIDR
    is remote Service ClusterIP CIDR, the tunnel IP in NXM_NX_TUN_IPV4_DST is remote Gateway IP):
    ```
    table=L3Forwarding, priority=200,ip,nw_dst=10.96.0.0/12 actions=mod_dl_src:ee:73:a5:81:09:c6,mod_dl_dst:aa:bb:cc:dd:ee:f0,load:0xab01b39->NXM_NX_TUN_IPV4_DST[],load:0x1->NXM_NX_REG0[4..7],resubmit(,L3DecTTL)
    ```
    * Add a rule in L3Forwarding table for cross-cluster reply packets. A sample is like below
      (the destination IP is remote Gateway IP):
    ```
    table=L3Forwarding, priority=200,ct_state=+rpl+trk,ip,nw_dst=10.176.27.57 actions=mod_dl_src:ee:73:a5:81:09:c6,mod_dl_dst:aa:bb:cc:dd:ee:f0,load:0xab01b39->NXM_NX_TUN_IPV4_DST[],load:0x1->NXM_NX_REG0[4..7],resubmit(,L3DecTTL)
    ```
    * Add a rule to do SNAT in Gateway. A sample is like below (the destination CIDR is remote
      Service ClusterIP CIDR, the SNAT source IP is local Gateway IP)::
    ```
    table=SNATConntrackCommit, priority=200,ip,nw_dst=10.96.0.0/12 actions=ct(commit,table=L2ForwardingCalc,zone=65530,exec(move:NXM_NX_REG0[0..3]->NXM_NX_CT_MARK[0..3],nat(src=10.176.25.59)))
    ```
    * Add a rule in L2ForwardingCalc table to load the global virtual multi-cluster MAC's output to `antrea-tun0`
    ```
    table=L2ForwardingCalc, priority=200,dl_dst=aa:bb:cc:dd:ee:f0 actions=load:0x1->NXM_NX_REG1[],load:0x1->NXM_NX_REG0[8],resubmit(,22)
    ```
    * Add a rule in ConntrackCommit table to match the request packets from regular Node and skip commit it.
    ```
    table=ConntrackCommit, priority=210,ct_mark=0x1/0xf,ip actions=resubmit(,Output)
    ```
    * Add a rule in Output table to match the multi-cluster traffic to forward the traffic from/to regular Node
    through the same port.
    ```
    table=Output, priority=210,reg0=0x200/0x200,reg1=0x1,in_port=1 actions=IN_PORT
    ```
* Add a controller for regular Nodes to watch Gateway and ClusterInfoImport's events.
   It will set up a few openflow rules to forward cross-cluster traffic to local Gateway Node.
    * Add a rule in L3Forwarding table for cross-cluster request packets, and
    modify the destination MAC to global multicluster virtual MAC. A sample is like below
    (the destination CIDR is remote Service ClusterIP CIDR, the tunnel IP in NXM_NX_TUN_IPV4_DST is local Gateway's Internal IP.):
    ```
    table=L3Forwarding, priority=200,ip,nw_dst=10.96.0.0/12 actions=mod_dl_src:f2:08:93:0c:82:bd,mod_dl_dst:aa:bb:cc:dd:ee:ff,load:0xab0193b->NXM_NX_TUN_IPV4_DST[],load:0x1->NXM_NX_REG0[4..7],resubmit(,L3DecTTL)
    ```
    * Add a rule in L3Forwarding table for cross-cluster reply packets. A sample is like below
    (the destination IP is remote Gateway IP, the tunnel IP in NXM_NX_TUN_IPV4_DST is local Gateway's Internal IP):
    ```
    table=L3Forwarding, priority=200,ct_state=+rpl+trk,ip,nw_dst=10.176.27.57 actions=mod_dl_src:f2:08:93:0c:82:bd,mod_dl_dst:aa:bb:cc:dd:ee:f0,load:0xab0193b->NXM_NX_TUN_IPV4_DST[],load:0x1->NXM_NX_REG0[4..7],resubmit(,L3DecTTL)
    ```
    * Add a rule in L2ForwardingCalc table to load the global virtual multi-cluster MAC's output to `antrea-tun0`
    ```
    table=L2ForwardingCalc, priority=200,dl_dst=aa:bb:cc:dd:ee:f0 actions=load:0x1->NXM_NX_REG1[],load:0x1->NXM_NX_REG0[8],resubmit(,22)
    ```
* Add unit test cases
* Refine e2e test for data plane change

Signed-off-by: Lan Luo <luola@vmware.com>

XinShuYang commented:
/test-multicluster-dataplane-e2e

XinShuYang commented:
/test-multicluster-dataplane-e2e

XinShuYang commented:
/test-multicluster-dataplane-e2e

luolanzone commented:
/test-multicluster-dataplane-e2e

luolanzone commented:
/test-multicluster-dataplane-e2e

luolanzone commented:
/test-multicluster-dataplane-e2e

luolanzone commented:
/test-multicluster-dataplane-e2e

luolanzone commented:
/test-multicluster-dataplane-e2e

luolanzone commented:
/test-multicluster-dataplane-e2e

luolanzone commented:
/test-multicluster-dataplane-e2e

luolanzone commented:
/test-multicluster-dataplane-e2e

luolanzone commented:
/test-multicluster-dataplane-e2e

luolanzone commented:
/test-multicluster-dataplane-e2e

luolanzone commented:
/test-multicluster-dataplane-e2e

luolanzone commented:
/test-multicluster-dataplane-e2e

luolanzone commented:
/test-multicluster-dataplane-e2e

luolanzone commented:
/test-multicluster-dataplane-e2e

luolanzone commented:
/test-multicluster-dataplane-e2e

luolanzone commented:
/test-multicluster-dataplane-e2e

luolanzone commented:
/test-multicluster-dataplane-e2e

luolanzone committed:
Use Service ClusterIPs as MC Service's Endpoints

Use Service ClusterIPs instead of Pod IPs as MC Service's Endpoints.
The ServiceExport controller will only watch ServiceExport and
Service events, and wrap Service's ClusterIPs into a new Endpoint kind of
ResourceExport.

Signed-off-by: Lan Luo <luola@vmware.com>

luolanzone commented:
/test-multicluster-dataplane-e2e

hjiajing commented:
/test-multicluster-e2e

hjiajing commented:
/test-multicluster-dataplane-e2e

luolanzone commented:
Closed this PR considering MC e2e need both commits in this PR and PR3603 to work, so cherry-picked this commit to PR3603.
